PT-2012 (PT-2785, PT-2777, PT-2944, PT-2929, PT-2942) - store genes as Ensembl IDs

allasm opened the pull request:
@a-gvo, @veronikaslc please review. (I'm not sure if @sashaandjic is back before the release, but someone needs to test this as well)

PT-2785 and PT-2777 have been reviewed already, cherry-picked and slightly modified in this branch and re-reviewed and approved by @a-gvo; thus only PT-2944, PT-2929 and PT-2942 need to be reviewed (and everything tested together)

veronikaslc commented on components/pedigree/resources/src/main/resources/pedigree/gene.js:
```diff
@@ -0,0 +1,104 @@
+        }
+    });
+
+    Gene.isValidEnsembleID = function(id) {
```
currently unused

allasm commented on components/pedigree/resources/src/main/resources/pedigree/view/person.js:
```diff
@@ -1571,5 +1565,30 @@ define([
+    Person.getGeneIndex = function(geneID, geneArray)
+    {
+        for (var i = 0; i < geneArray.length; i++) {
+            if (geneArray[i].hasOwnProperty("id")) {
```
because the other "else" should happen only if `!geneArray[i].hasOwnProperty("id")`, but not when `geneArray[i].hasOwnProperty("id")` and `geneArray[i].id.toUpperCase() != geneID.toUpperCase()`

allasm commented:
addressed all comments and rebased on top of master

@sashaandjic - there were a few minor conflicts in UI code related to patient form, if you can use this new version and try playing with genes on the patient page that would be great

Also need to test migration from an older version which has pedigree with genes

sashaandjic commented:
Now I can enter and save the same gene 3 times (casual, candidate and negative).

AlexLapena commented:
I can also enter the same gene multiple times.

AlexLapena commented:
Pushing genes from 1.2.5 -> PT-2012 does not work.

sashaandjic commented:
@allasm  @sdumitriu @AlexLapena @marta- The bug that allows saving same genes 
multiple times is present in the master. 
![screen shot 2016-12-08 at 3 28 55 pm](https://cloud.githubusercontent.com/assets/23243244/21026657/a2e91904-bd5b-11e6-8701-68a6707eca7c.png)

allasm commented:
(rebased on top of latest master to resolve a conflict)

@AlexLapena: 1.2.x can not push genes (my bad if I somewhere suggested it should be tested), so what actually need to be tested is exporting JSON in 1.2.x and importing in this version (can do via new conveninet `import JSON` menu item without having to use `curl`)

@a-gvo and/or @veronikaslc - can you please look at the patient form gene-validation code, according to @marta- it used to validate and not allow the same gene to be entered multiple times, but not it allows that. According to Marta it is possible some of the @sdumitriu changes caused this, b we need to fix this anyhow

veronikaslc commented:
I will look into it.

veronikaslc commented:
Made a [PR](https://github.com/phenotips/phenotips/pull/2175/files) to master with the gene validation regression fix.

allasm commented:
Rebased on master (to fix duplicate gene problem) and fixed how duplicate genes are handled in 1.2.x JSONs. Minor code review is required, and a small re-testing of JSON importing for both 1.2.x and 1.3.x JSONs (@sashaandjic)

sashaandjic commented:
found problem with PED export when selecting specific gene(s); every node is marked as affected (-9)

veronikaslc commented:
build worked for me. so +1.

allasm commented:
Fixed PED export when genes are selected. It was a 2 line fix which can only possibly affect Export, so that is the only thing to (re-)test

sashaandjic commented:
found another issue, not a show stopper though. here are the steps:
- create new patient
- open pedigree editor and import the following simple JSON 
[{"id":"1","proband":true,"father":"3","mother":"2","firstName":"AMELY","sex":"unknown","birthDate":{},"deathDate":{},"lifeStatus":"alive","externalId":"","genes":[{"gene":"AMELY","id":"ENSG00000099721","status":"candidate"}]},{"id":"2","lastName":" + Type C brachydactyly","features":[{"id":"HP:0009373","label":"Type C brachydactyly","type":"phenotype","observed":"yes"}],"firstName":"BCS1L","sex":"female","genes":[{"gene":"BCS1L","id":"ENSG00000074582","status":"solved"}],"nonstandard_features":[]},{"id":"3","features":[],"firstName":"BCS1L","sex":"male","genes":[{"gene":"BCS1L","id":"ENSG00000074582","status":"solved"}],"nonstandard_features":[]}] 

- drag the patient node to the proband node called AMELY (it has AMELY as candidate gene selected)
- notice that the node colour and name disappears
 - hover over AMELY gene on the right panel, the correct node is still highlighted
- click on the node and notice that there is no AMELY candidate gene entry in there

allasm commented:
- fixed PED export when genes are selected. It was a 2 line fix which can only possibly affect Export, so that is the only thing to (re-)test for this change

- fixed patient dragging bug. Need to re-test what happens when patient is dragged to a node with genes and phenotypes and disorders, if the patient had some before or had none before. Have to test what happens if a new patient is created for a node which has genes and phenotypes and disorders. Draged both after import and after unlinking

sashaandjic commented:
both defects fixed +1
